torch/csrc/jit/tensorexpr/kernel.cpp

```diff
@@ -2060,6 +2084,42 @@ Tensor* TensorExprKernel::computeConv2d(const torch::jit::Value* v) {
   return new Tensor(ResultBuf.node(), s);
 }
 
+Tensor* TensorExprKernel::computeMatmul(const torch::jit::Value* v) {
+  const Node* n = v->node();
+  auto const& shape = sizesForValue(v);
+  Dtype dtype = kFloat;
+  auto maybe_stype = findDtypeForValue(v);
+  if (maybe_stype) {
+    dtype = Dtype(*maybe_stype);
+  }
+  BufHandle ResultBuf("matmul", shape, dtype);
+  const Buf* a = tensors_.at(n->input(0))->buf();
+  const Buf* b = tensors_.at(n->input(1))->buf();
+
+  auto size_a = ExprVectorToExprHandleVector(a->dims());
+  auto size_b = ExprVectorToExprHandleVector(b->dims());
+  const IntImm* total_size = dynamic_cast<const IntImm*>(
+      IRSimplifier::simplify((size_a[0] * size_a[1] * size_b[1])).node());
+
+  if (total_size && total_size->value() < 1000) {
+    return Reduce(
+        "nnc_matmul",
+        {{size_a[0], "M"}, {size_b[1], "N"}},
+        Sum(),
+        [&](const ExprHandle& m, const ExprHandle& n, const ExprHandle& k) {
+          BufHandle ah(a);
+          BufHandle bh(b);
+          return Load::make(ah, {m, k}) * Load::make(bh, {k, n});
+        },
+        {{size_a[1], "K"}});
```

Comment:
In the spirit of https://randomascii.wordpress.com/2014/01/27/theres-only-four-billion-floatsso-test-them-all/: there are a very small number of possible inputs to test, you might try writing a script to see how large you can make the total size while nnc is as fast as aten matmul (with a little bit of a leeway for possible fusion, say 5%)

Reply:
This is a very basic naive heuristic and we for sure would like to tune it (and hopefully add better schedules for matmul as well). But I think it deserves a separate PR too, this one is mostly to lay down the foundation for this work.